### PR TITLE
AFA data structures, Closed set data structures, functions for inspecting an AFA, two AFA emptiness tests

### PR DESCRIPTION
**Pull request by kocotom**
I've changed the former data structure for alternating automata, I've added structures for transition and inverse transition and several functions which enable us to inspect an alternating automaton using antichains (pre, post). I've also implemented two antichain-based AFA emptiness tests (which work in a forward and backward fashion).
I've also added a file closed_set.hh which makes the work with antichains easier.
I've also added several test cases.

### CODE REVIEW

File: src/afa/afa.cc
Comment (tfiedor) [inline src/afa/afa.cc:481]: Is this really the efficient process? It seams shady that you perform the transition to check if the automaton has the transition. Can you elaborate?

File: src/afa/afa.cc
Comment (tfiedor) [inline src/afa/afa.cc:758]: Please, don't keep commented code in the PR.

**Review by tfiedor — APPROVED** [review]
I generally approve of this PR, the code is readable, and commented. I'm not AFA-expert so I cannot fully assert the correctness itself, though.

There are some areas that could be improved, and I advise to at least consider:

1. While there are comments, I'd say they could be more clear for non-AFA experts
2. Some naming could be improved (the classes/variables with `Inv` prefix)
3. I would improve the asserts by adding string explaining the failure, e.g. `assert(!trans.empty() && "Transition Relation has to be allocated first")
4. On several places you call function in for-loop condition, which can lead to performance problems.

File: include/mata/closed_set.hh
Comment (Adda0) [reply include/mata/closed_set.hh:207]: Do you mean the capital `U`? You cannot use the name `union` as that is a [keyword in C++](https://en.cppreference.com/w/cpp/language/union). We use `Union` elsewhere in the code already, too.

File: src/afa/tests-afa.cc
Comment (Adda0) [inline src/afa/tests-afa.cc:267]: We need to discuss this, in my opinion. Otherwise, this comment will be forgotten and left unresolved.

File: src/afa/tests-afa.cc
Comment (Adda0) [inline src/afa/tests-afa.cc:360]: Either remove this test, or fix it, but do not keep this test commented out just like this.

And if you decide to remove the test, @martinhruska should know it and be aware that the parsing of AFAs no longer works. Do I understand it right?

File: src/afa/afa.cc
Comment (Adda0) [inline src/afa/afa.cc:71]: Discuss this with others. We should decide on this before we merge the PR.

File: src/afa/afa.cc
Comment (Adda0) [inline src/afa/afa.cc:120]: Looking at the use of `sharing_list` now, I really do not like the name. I have no idea from the name itself, what is it supposed to do. Could we maybe come with a more appropriate name?

File: src/afa/afa.cc
Comment (Adda0) [inline src/afa/afa.cc:98]: Could we try to keep the code contained within the 120 characters boundary? I had to scroll to the right in my IDE to read the whole comment, which is really unpleasant.

File: src/afa/afa.cc
Comment (Adda0) [inline src/afa/afa.cc:362]: Am I right?
```suggestion
* @brief performs an inverse transition using a single state and symbol
```

File: src/afa/afa.cc
Comment (Adda0) [reply src/afa/afa.cc:758]: And if there is some TODO, explain what is actually there to do.

File: src/afa/afa.cc
Comment (Adda0) [inline src/afa/afa.cc:876]: Again here. What are we supposed to do? Why to keep the commented out line here? Try to resolve it.

**Review by Adda0 — CHANGES_REQUESTED** [review]
Looks good to me in general, although I cannot say I understand AFAs and all of the introduced operations as much as I would like to. Therefore, I cannot confirm the correctness of your code, either.

My views on this PR are pretty much the same as @tfiedor's. Try to improve comments wherever possible (make them more clear, resolve ambiguousness, etc.), improve naming and resolve obvious issues such as the commented out code, some of the TODOs, etc. Please, feel free to discuss the feedback whenever you do not agree with the suggested changes.

Otherwise, I think the implementation is sound, and we have agreed to go with your design of used structures and classes to see how it behaves on some benchmarks. If we find out the design is unfit for our uses, we can try to come up with something else.

File: include/mata/afa.hh
Comment (kocotom) [reply include/mata/afa.hh:75]: Thanks for your suggestion. I've overlooked it. I will be more careful next time.

File: src/afa/afa.cc
Comment (kocotom) [reply src/afa/afa.cc:362]: Yes, that is a mistake, thanks for your suggestion!

File: include/mata/afa.hh
Comment (martinhruska) [reply include/mata/afa.hh:70]: Yes it would be nice to define what nodes are in retion to standard definition of AFA.

File: include/mata/afa.hh
Comment (martinhruska) [inline include/mata/afa.hh:101]: The comments such this are very technical. It basically describes implementation but I would appreciate generally describe what InvResults are and then describe how this particular data structure represents it.

**Review by martinhruska — COMMENTED** [review]
Everything important was already said by other reviewers. I added only minor comments. The mentioned issues should be resolved before PR is going to be merged.

**Comment by martinhruska** [comment]
So are all comments addressed? Could you mark conversations as resolved? Once everything is done, we can merge this PR.

File: include/mata/afa.hh
Comment (kocotom) [reply include/mata/afa.hh]: Thanks for your suggestion. I have renamed the structure to make it more understandable and I have also changed the comment.

File: include/mata/afa.hh
Comment (kocotom) [reply include/mata/afa.hh]: Thanks for your suggestion. I have renamed the structure to make it more understandable and I also have changed the comment.

File: include/mata/afa.hh
Comment (kocotom) [reply include/mata/afa.hh]: Thanks for your suggestion. I was thinking about these structures a lot and I realised that no pointers are necessary here.
Initially, I thought I needed to store each `InvResults` to the inverse transition relation multiple times to make it accessible from all the states of the corresponding `sharing_list`. Therefore I decided to use pointers to get rid of redundancy and avoid searching for all occurrences of the corresponding `InvResult` in case of changing its value. 
Then, I realised that I could choose one representative value of the `sharing_list` (minimal element of the set) and make the `InvResult` accessible in the transition relation only via this element so no pointers are needed anymore.

File: include/mata/afa.hh
Comment (kocotom) [reply include/mata/afa.hh]: Thanks for your suggestion. I have renamed the structure to make it more understandable and unambiguous.

File: include/mata/afa.hh
Comment (kocotom) [reply include/mata/afa.hh]: Thanks for your suggestion. I have added another implementation of this function so by now, there are two functions with even longer names `antichain_concrete_forward_emptiness_test_old` and `antichain_concrete_forward_emptiness_test_new` which will be compared. 
I agree that we should come up with better names.

File: include/mata/closed_set.hh
Comment (kocotom) [reply include/mata/closed_set.hh]: Thanks for your suggestion. You are right, `static` was not necessary there, I have fixed it.

File: include/mata/closed_set.hh
Comment (kocotom) [reply include/mata/closed_set.hh]: Thanks for your suggestion, I have reimplemented it using `enum`.

File: include/mata/closed_set.hh
Comment (kocotom) [reply include/mata/closed_set.hh]: Thanks for your suggestion, I have removed it.

File: include/mata/closed_set.hh
Comment (kocotom) [reply include/mata/closed_set.hh]: Thanks for your suggestion, I have reimplemented it.

File: include/mata/closed_set.hh
Comment (kocotom) [reply include/mata/closed_set.hh]: I hope it is ok, I need to check whether all states which were stored to the `node` respect the interval `<min_val, max_val>` of the closed set.

File: src/afa/afa.cc
Comment (kocotom) [reply src/afa/afa.cc]: Thanks for your suggestion, the assertion did not make any sense. I have removed it.

File: src/afa/afa.cc
Comment (kocotom) [reply src/afa/afa.cc]: Thanks for your suggestion. This function was completely removed in the current version of the code but I have added comments to the rest of the assertions.

File: src/afa/afa.cc
Comment (kocotom) [reply src/afa/afa.cc]: Thanks for your suggestion, you are right, it was not necessary. However, I have completely removed this function in the latest version of the code.

File: src/afa/afa.cc
Comment (kocotom) [reply src/afa/afa.cc]: Thanks for your suggestion, I have reimplemented it.

File: src/afa/afa.cc
Comment (kocotom) [reply src/afa/afa.cc]: Thanks for your suggestion, I have overlooked it. The comment is removed now.

File: src/afa/afa.cc
Comment (kocotom) [reply src/afa/afa.cc]: Thanks for your suggestion, I think this is a nice idea so I have implemented it.

File: include/mata/afa.hh
Comment (kocotom) [reply include/mata/afa.hh]: Thanks for your suggestion, I have rewritten it to provide more details.

File: include/mata/afa.hh
Comment (kocotom) [reply include/mata/afa.hh]: Thanks for your suggestion, I have removed it.

File: include/mata/afa.hh
Comment (kocotom) [reply include/mata/afa.hh]: Thanks for your suggestion! I have renamed these structures to make them more understandable.
Thanks for the note about the keyword `typedef`, I had no idea I could remove it. It is removed in the latest version of the code.

File: src/afa/afa.cc
Comment (kocotom) [reply src/afa/afa.cc]: Thanks for your suggestion, I have stressed it in the latest version of the code.

File: src/afa/afa.cc
Comment (kocotom) [reply src/afa/afa.cc]: Thanks for your suggestion, I have written this as a note for myself and I completely forgot to remove it before I created the pull request. I will be more careful next time.

File: include/mata/afa.hh
Comment (kocotom) [reply include/mata/afa.hh]: Thanks for your suggestion, I did not notice that the former code used `TAB` for indentation while I have used multiple spaces. The difference was not apparent in my editor. I have fixed it and I will be more careful next time.

File: src/afa/tests-afa.cc
Comment (kocotom) [reply src/afa/tests-afa.cc]: Thanks for your suggestion, I have removed the comment and I have discussed it with others.

File: src/afa/tests-afa.cc
Comment (kocotom) [reply src/afa/tests-afa.cc:432]: Thanks for your suggestion, I asked @martinhruska and he said it is ok to leave it commented for now while he is working on the parsing of AFAs.

File: src/afa/afa.cc
Comment (kocotom) [reply src/afa/afa.cc]: Thanks for your suggestion, we intend to discuss it.

File: src/afa/afa.cc
Comment (kocotom) [reply src/afa/afa.cc]: Thanks for your suggestion, I did not realise. I have fixed it.

File: src/afa/afa.cc
Comment (kocotom) [reply src/afa/afa.cc]: Thanks for your suggestion, I have described the issue in the latest version of the code.

File: src/afa/afa.cc
Comment (kocotom) [reply src/afa/afa.cc]: Thanks for your suggestion, I have reimplemented it.
It was possible to easily change the function `get_final_nodes()` since the corresponding antichain has a format `{{f_1, f_2, f_3, ... f_n}}`, where `f_1, f_2, ... f_n` are final states. However, the function `get_initial_nodes()` needs to create an antichain which has a format `{{s_1}, {s_2}, {s_3}, ... {s_n}}`, where `s_1, s_2, ... s_n` are initial states so I need to create several nodes and I cannot easily use the `initialstates` set.

File: src/afa/afa.cc
Comment (kocotom) [reply src/afa/afa.cc]: Thanks for your suggestion, we have discussed it. In the latest version of the code, I use the name `precondition`.

File: include/mata/afa.hh
Comment (martinhruska) [reply include/mata/afa.hh]: I will rewrite comments later.

**Comment by martinhruska** [comment]
It looks that the most comments by reviewers were addressed. @Adda0 are all requested changes done?

The last thing left for discussion is number of Post functions. However, I don't think it is possible to easily rewrite code without them. So I am for merging this PR and maybe creating issue about possible improvements of AFA code.

**Review by Adda0 — APPROVED** [review]
I agree with @martinhruska. I have checked the changes, and I am OK with merging this PR as of now. Eventually, we will resolve the mentioned issues. Good job.

**Pull request merged**
